src/plugins/intel_cpu/src/utils/serialize.cpp

```diff
@@ -0,0 +1,199 @@
+// Copyright (C) 2018-2024 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include "serialize.hpp"
+
+#include "openvino/core/descriptor_tensor.hpp"
+#include "openvino/core/parallel.hpp"
+#include "openvino/runtime/shared_buffer.hpp"
+
+namespace ov {
+namespace intel_cpu {
+
+inline void codec_xor(char* dst_str, const char* src_str, size_t len) {
+    static const char codec_key[] = {0x30, 0x60, 0x70, 0x02, 0x04, 0x08, 0x3F, 0x6F, 0x72, 0x74, 0x78, 0x7F};
+    auto key_size = sizeof(codec_key);
+
+    if (dst_str == src_str) {
+        parallel_for(len, [&](size_t key_idx) {
+            dst_str[key_idx] ^= codec_key[key_idx % key_size];
+        });
+    } else {
+        parallel_for(len, [&](size_t key_idx) {
+            dst_str[key_idx] = src_str[key_idx] ^ codec_key[key_idx % key_size];
+        });
+    }
+}
+
+std::string codec_xor_str(const std::string& source_str) {
+    std::string new_str(source_str);
+    codec_xor(&new_str[0], &new_str[0], new_str.size());
+    return new_str;
+}
+
+////////// ModelSerializer //////////
+
+ModelSerializer::ModelSerializer(std::ostream& ostream) : m_ostream(ostream) {}
+
+void ModelSerializer::operator<<(const std::shared_ptr<ov::Model>& model) {
+    auto serialize_info = [&](std::ostream& stream) {
+        pugi::xml_document xml_doc;
+        pugi::xml_node root = xml_doc.append_child("cnndata");
+        pugi::xml_node outputs = root.append_child("outputs");
+        for (const auto& out : model->get_results()) {
+            auto out_node = outputs.append_child("out");
+            const auto name = ov::descriptor::get_ov_tensor_legacy_name(out->input_value(0).get_tensor());
+            out_node.append_attribute("name").set_value(name.c_str());
+        }
+        xml_doc.save(stream);
+    };
+
+    ov::pass::StreamSerialize serializer(m_ostream, serialize_info, codec_xor_str);
+    serializer.run_on_model(std::const_pointer_cast<ov::Model>(model->clone()));
+}
+
+////////// ModelDeserializer //////////
+
+ModelDeserializer::ModelDeserializer(std::istream& model_stream, ModelBuilder fn)
+    : m_istream(model_stream), m_model_builder(fn) {}
+
+void ModelDeserializer::set_info(pugi::xml_node& root, std::shared_ptr<ov::Model>& model) {
+    pugi::xml_node outputs = root.child("outputs");
+    auto nodes_it = outputs.children("out").begin();
+    size_t size = model->outputs().size();
+    for (size_t i = 0lu; i < size; ++nodes_it, i++) {
+        std::string name = nodes_it->attribute("name").value();
+        if (name.empty())
+            continue;
+        auto result = model->output(i).get_node_shared_ptr();
+        ov::descriptor::set_ov_tensor_legacy_name(result->input_value(0).get_tensor(), name);
+    }
+}
+
+void ModelDeserializer::operator>>(std::shared_ptr<ov::Model>& model) {
+    if (auto stream_buf = dynamic_cast<MmapStreamBuffer*>(m_istream.rdbuf())) {
```

Comment:
is it possible for users to disable mmap in import_model / model cache ?
E.g. they want to clean up cache from time to time and if we refer to cache files via mmap, such files cannot be removed on Windows.

Reply:
No